File: infra/services/clogs/snorkel.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"reflect"
)

const (
	IntType = 1
	StrType = 2
	SetType = 3
)

type EventData map[string]interface{}

type event struct {
	Table string    `json:"table"`
	Token string    `json:"token"`
	Data  EventData `json:"data"`
}

type Snorkel struct {
	table string
	token string
	spec  map[string]int
}

func NewSnorkel(table string, token string) *Snorkel {
	return &Snorkel{
		table: table,
		token: token,
		spec:  map[string]int{},
	}
}

func (s *Snorkel) AddIntField(name string) *Snorkel {
	return s.addField(name, IntType)
}

func (s *Snorkel) AddStrField(name string) *Snorkel {
	return s.addField(name, StrType)
}

func (s *Snorkel) AddSetField(name string) *Snorkel {
	return s.addField(name, SetType)
}

func (s *Snorkel) addField(name string, fieldType int) *Snorkel {
	s.spec[name] = fieldType
	return s
}

func (s *Snorkel) Write(writer io.Writer, data map[string]interface{}) error {
	out := map[string]interface{}{}
	for k, v := range data {
		ftype, ok := s.spec[k]
		if !ok {
			continue
		}
		switch ftype {
		case IntType:
			intv, err := toInt(v)
			if err != nil {
				return fmt.Errorf("int type expected, key=%v, value=%v, %w", k, v, err)
			}
			out[k] = intv
		case StrType:
			_, ok := v.(string)
			if !ok {
				return fmt.Errorf("string type expected, key=%v, value=%v", k, v)
			}
			out[k] = v
		case SetType:
			_, ok := v.([]string)
			if !ok {
				return fmt.Errorf("expected an array of string type, but key=%v, value=%v", k, v)
			}
			out[k] = v
		default:
			return fmt.Errorf("unknown type is given, type=%v", ftype)
		}
	}

	evt := event{
		Table: s.table,
		Token: s.token,
		Data:  out,
	}

	bs := toJson(evt)
	bsln := append(bs, []byte("\n")...)
	if nwritten, err := writer.Write(bsln); err != nil {
		return err
	} else if nwritten != len(bs) {
		return fmt.Errorf(
			"failed to write all data, written=%v, expected=%v",
			nwritten,
			len(bs))
	}

	return nil
}

func toJson(e event) []byte {
	bs, err := json.Marshal(e)
	if err != nil {
		panic(err)
	}
	return bs
}

func toInt(v interface{}) (int64, error) {
	if v == nil {
		return 0, errors.New("not compatiable")
	}

	rv := reflect.ValueOf(v)
	rt := rv.Type()
	if !rt.ConvertibleTo(reflect.TypeOf(int64(0))) {
		return 0, errors.New("not compatiable")
	}
	value := rv.Convert(reflect.TypeOf(int64(0)))
	return value.Int(), nil
}

File: prod/xplat/snorkel-py/snorkel.py
from enum import Enum
from typing import Dict, Optional
import json


class FieldType(Enum):
    INT = 1
    STR = 2
    SET = 3


class Snorkel:

    def __init__(self, table: str, token: str) -> None:
        self.table = table
        self.token = token
        self.spec = {}

    def add_int_field(self, name: str):
        self.__add_field(name, FieldType.INT)

    def add_str_field(self, name: str):
        self.__add_field(name, FieldType.STR)

    def add_set_field(self, name: str):
        self.__add_field(name, FieldType.SET)

    def __add_field(self, name: str, ftype: FieldType):
        self.spec[name] = ftype

    def write(self, data: dict[str, int | str | list[str]]):
        out = {}
        for k, v in data.items():
            if not k in self.spec:
                continue

            ftype = self.spec[k]
            if ftype == FieldType.INT:
                intv = Snorkel.to_int(v)
                if intv is None:
                    raise ValueError(
                        f'a value is not compatible with int, k={k}')
                out[k] = intv
            elif ftype == FieldType.STR:
                strv = f'{v}'
                out[k] = strv
            elif ftype == FieldType.SET:
                if not isinstance(v, list):
                    raise ValueError(f'expected a list of string, k={k}')
                out[k] = [f'{x}' for x in list]
            else:
                raise ValueError(f'an unknown type is given, k={k}')

        evt = {'table': self.table, 'token': self.token, 'data': out}
        evt_json = json.dumps(evt)
        print(f'{evt_json}\n')

    @staticmethod
    def to_int(v: int | str | list[str]) -> int | None:
        t = type(v)
        if t == int or t == float:
            return int(v)
        if t == str:
            try:
                return int(v)
            except Exception:
                return None

        return None
